numpy/array_api/_typing.py

```diff
@@ -33,12 +35,26 @@
 
 # This should really be recursive, but that isn't supported yet. See the
 # similar comment in numpy/typing/_array_like.py
-NestedSequence = Sequence[Sequence[Any]]
+_T = TypeVar("_T")
+NestedSequence = Sequence[Sequence[_T]]
```

Comment:
`NestedSequence` needs a free (subscriptable) parameter based on how it's used in `np.array_api.asarray`:
https://github.com/numpy/numpy/blob/2d112a98ed7597c4120b31908384ae09b0304659/numpy/array_api/_creation_functions.py#L38

Reply:
Off topic: mypy has some basic support for recursive types these days,
so it is now possible to define a proper arbitrary-nested sequence type with the help of protocols (xref https://github.com/numpy/numpy/pull/19894)